File: employee_scheduler.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <algorithm>
#include <random>
#include <iomanip>

using namespace std;

// ============================================================
// Assignment 4: Implementing Control Structures
// Employee Shift Scheduler - C++ Version
// ============================================================

// Constants
const int DAYS = 7;
const int SHIFTS = 3;
const int MIN_EMPLOYEES = 9;
const int MIN_PER_SHIFT = 2;
const int MAX_DAYS_PER_EMPLOYEE = 5;

// Days of the week
const vector<string> dayNames = {
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday",
    "Sunday"
};

// Available shifts
const vector<string> shiftNames = {
    "Morning",
    "Afternoon",
    "Evening"
};


// ============================================================
// Employee Structure
// ============================================================

struct Employee
{
    string name;

    // Preference for each day:
    // 0 = Morning
    // 1 = Afternoon
    // 2 = Evening
    // -1 = No Preference
    vector<int> preferences;

    int daysWorked;
};


// ============================================================
// Schedule Structure
// ============================================================

struct Schedule
{
    // schedule[day][shift] contains employee indexes
    vector<vector<vector<int>>> assignments;

    // Records when an employee's preferred shift was unavailable
    vector<string> conflicts;
};


// ============================================================
// Random Number Generator
// ============================================================

random_device rd;
mt19937 generator(rd());


// ============================================================
// Convert Shift Number to Text
// ============================================================

string shiftToString(int shift)
{
    if (shift == 0)
    {
        return "Morning";
    }
    else if (shift == 1)
    {
        return "Afternoon";
    }
    else
    {
        return "Evening";
    }
}


// ============================================================
// Convert User Input to Shift Number
// ============================================================

int getShiftChoice()
{
    int choice;

    while (true)
    {
        cout << "    1. Morning";
        cout << "    2. Afternoon";
        cout << "    3. Evening";
        cout << "    4. No Preference\n";

        cout << "    Enter choice: ";
        cin >> choice;

        if (choice >= 1 && choice <= 4)
        {
            if (choice == 4)
            {
                return -1;
            }

            return choice - 1;
        }

        cout << "    Invalid choice. Please enter 1, 2, 3, or 4.\n";
    }
}


// ============================================================
// Check for Duplicate Employee Name
// ============================================================

bool duplicateName(
    const vector<Employee>& employees,
    const string& name)
{
    for (const Employee& employee : employees)
    {
        if (employee.name == name)
        {
            return true;
        }
    }

    return false;
}


// ============================================================
// Input Employee Information
// ============================================================

void inputEmployees(vector<Employee>& employees)
{
    int numberOfEmployees;

    cout << "\n";
    cout << "============================================================\n";
    cout << "              EMPLOYEE INFORMATION\n";
    cout << "============================================================\n";

    cout << "\nEnter number of employees (9-12): ";
    cin >> numberOfEmployees;

    while (numberOfEmployees < MIN_EMPLOYEES ||
           numberOfEmployees > 12)
    {
        cout << "\nPlease enter between 9 and 12 employees.\n";
        cout << "Enter number of employees: ";
        cin >> numberOfEmployees;
    }

    cin.ignore();

    for (int i = 0; i < numberOfEmployees; i++)
    {
        Employee employee;

        cout << "\n------------------------------------------------------------\n";
        cout << "Employee " << i + 1 << "\n";
        cout << "------------------------------------------------------------\n";

        cout << "Enter employee name: ";
        getline(cin, employee.name);

        while (employee.name.empty() ||
               duplicateName(employees, employee.name))
        {
            if (employee.name.empty())
            {
                cout << "Name cannot be empty.\n";
            }
            else
            {
                cout << "That employee name already exists.\n";
            }

            cout << "Enter employee name: ";
            getline(cin, employee.name);
        }

        employee.preferences.resize(DAYS);
        employee.daysWorked = 0;

        cout << "\nEnter shift preferences for each day.\n";
        cout << "You can select Morning, Afternoon, Evening, or No Preference.\n";

        for (int day = 0; day < DAYS; day++)
        {
            cout << "\n" << dayNames[day] << ":\n";

            employee.preferences[day] = getShiftChoice();
        }

        employees.push_back(employee);
    }
}


// ============================================================
// Check Whether Employee Is Working on a Specific Day
// ============================================================

bool worksOnDay(
    const Schedule& schedule,
    int day,
    int employeeIndex)
{
    for (int shift = 0; shift < SHIFTS; shift++)
    {
        for (int employee : schedule.assignments[day][shift])
        {
            if (employee == employeeIndex)
            {
                return true;
            }
        }
    }

    return false;
}


// ============================================================
// Count Employees Assigned to a Shift
// ============================================================

int shiftCount(
    const Schedule& schedule,
    int day,
    int shift)
{
    return static_cast<int>(
        schedule.assignments[day][shift].size()
    );
}


// ============================================================
// Determine Whether a Shift Is Full
// ============================================================

bool shiftIsFull(
    const Schedule& schedule,
    int day,
    int shift)
{
    return shiftCount(schedule, day, shift) >= MIN_PER_SHIFT;
}


// ============================================================
// Calculate Preference Score
// ============================================================

int preferenceScore(
    const Employee& employee,
    int day,
    int shift)
{
    int preferredShift = employee.preferences[day];

    if (preferredShift == shift)
    {
        return 100;
    }

    if (preferredShift == -1)
    {
        return 50;
    }

    // Employee prefers another shift.
    return 0;
}


// ============================================================
// Select Six Employees for Each Day
//
// Six employees are needed each day because:
// 3 shifts x 2 employees = 6 assignments per day.
//
// Employees with fewer previous workdays are prioritized.
// This helps ensure that nobody exceeds five days.
// ============================================================

vector<int> selectEmployeesForDay(
    vector<Employee>& employees,
    int day)
{
    vector<int> candidates;

    for (int i = 0; i < static_cast<int>(employees.size()); i++)
    {
        candidates.push_back(i);
    }

    // Randomize equal-priority candidates.
    shuffle(
        candidates.begin(),
        candidates.end(),
        generator
    );

    // Sort primarily by number of days worked.
    sort(
        candidates.begin(),
        candidates.end(),
        [&](int a, int b)
        {
            return employees[a].daysWorked <
                   employees[b].daysWorked;
        }
    );

    // Select six employees.
    vector<int> selected;

    for (int employeeIndex : candidates)
    {
        if (employees[employeeIndex].daysWorked <
            MAX_DAYS_PER_EMPLOYEE)
        {
            selected.push_back(employeeIndex);
        }

        if (selected.size() == 6)
        {
            break;
        }
    }

    return selected;
}


// ============================================================
// Generate Best Shift Assignment for Six Employees
//
// There are six employees and six positions:
//
// Morning:   2
// Afternoon: 2
// Evening:   2
//
// We test possible arrangements and choose the arrangement
// that satisfies the largest number of preferences.
// ============================================================

vector<int> bestShiftAssignment(
    const vector<Employee>& employees,
    const vector<int>& selected,
    int day)
{
    vector<int> bestAssignment;
    int bestScore = -1;

    vector<int> permutation = selected;

    sort(
        permutation.begin(),
        permutation.end()
    );

    do
    {
        int score = 0;

        // First two employees -> Morning
        for (int i = 0; i < 2; i++)
        {
            score += preferenceScore(
                employees[permutation[i]],
                day,
                0
            );
        }

        // Next two -> Afternoon
        for (int i = 2; i < 4; i++)
        {
            score += preferenceScore(
                employees[permutation[i]],
                day,
                1
            );
        }

        // Last two -> Evening
        for (int i = 4; i < 6; i++)
        {
            score += preferenceScore(
                employees[permutation[i]],
                day,
                2
            );
        }

        if (score > bestScore)
        {
            bestScore = score;
            bestAssignment = permutation;
        }

    } while (next_permutation(
        permutation.begin(),
        permutation.end()
    ));

    return bestAssignment;
}


// ============================================================
// Generate Weekly Schedule
// ============================================================

Schedule generateSchedule(
    vector<Employee>& employees)
{
    Schedule schedule;

    schedule.assignments.resize(DAYS);

    for (int day = 0; day < DAYS; day++)
    {
        schedule.assignments[day].resize(SHIFTS);

        // Select six employees for this day.
        vector<int> selected =
            selectEmployeesForDay(
                employees,
                day
            );

        if (selected.size() < 6)
        {
            continue;
        }

        // Find the assignment that best matches preferences.
        vector<int> bestAssignment =
            bestShiftAssignment(
                employees,
                selected,
                day
            );

        // Assign two employees to each shift.
        for (int position = 0; position < 6; position++)
        {
            int employeeIndex =
                bestAssignment[position];

            int shift = position / 2;

            schedule.assignments[day][shift]
                .push_back(employeeIndex);

            employees[employeeIndex].daysWorked++;
        }
    }

    // ========================================================
    // Detect and Record Preference Conflicts
    // ========================================================

    for (int day = 0; day < DAYS; day++)
    {
        for (int shift = 0; shift < SHIFTS; shift++)
        {
            for (int employeeIndex :
                 schedule.assignments[day][shift])
            {
                int preferred =
                    employees[employeeIndex]
                    .preferences[day];

                if (preferred != -1 &&
                    preferred != shift)
                {
                    string conflict =
                        employees[employeeIndex].name +
                        " preferred " +
                        shiftToString(preferred) +
                        " on " +
                        dayNames[day] +
                        " but was assigned to " +
                        shiftToString(shift) +
                        ".";

                    schedule.conflicts.push_back(
                        conflict
                    );
                }
            }
        }
    }

    return schedule;
}


// ============================================================
// Validate Minimum Staffing
// ============================================================

bool validateMinimumStaffing(
    const Schedule& schedule)
{
    for (int day = 0; day < DAYS; day++)
    {
        for (int shift = 0; shift < SHIFTS; shift++)
        {
            if (shiftCount(
                    schedule,
                    day,
                    shift) < MIN_PER_SHIFT)
            {
                return false;
            }
        }
    }

    return true;
}


// ============================================================
// Validate One Shift Per Day
// ============================================================

bool validateOneShiftPerDay(
    const Schedule& schedule,
    int employeeCount)
{
    for (int day = 0; day < DAYS; day++)
    {
        for (int employee = 0;
             employee < employeeCount;
             employee++)
        {
            int numberOfShifts = 0;

            for (int shift = 0;
                 shift < SHIFTS;
                 shift++)
            {
                for (int assignedEmployee :
                     schedule.assignments[day][shift])
                {
                    if (assignedEmployee == employee)
                    {
                        numberOfShifts++;
                    }
                }
            }

            if (numberOfShifts > 1)
            {
                return false;
            }
        }
    }

    return true;
}


// ============================================================
// Validate Maximum Five Days
// ============================================================

bool validateMaximumFiveDays(
    const vector<Employee>& employees)
{
    for (const Employee& employee : employees)
    {
        if (employee.daysWorked >
            MAX_DAYS_PER_EMPLOYEE)
        {
            return false;
        }
    }

    return true;
}


// ============================================================
// Display Weekly Schedule
// ============================================================

void displaySchedule(
    const Schedule& schedule,
    const vector<Employee>& employees)
{
    cout << "\n";
    cout << "============================================================\n";
    cout << "                 FINAL WEEKLY SCHEDULE\n";
    cout << "============================================================\n";

    for (int day = 0; day < DAYS; day++)
    {
        cout << "\n";
        cout << "------------------------------------------------------------\n";
        cout << dayNames[day] << "\n";
        cout << "------------------------------------------------------------\n";

        for (int shift = 0; shift < SHIFTS; shift++)
        {
            cout << left
                 << setw(12)
                 << shiftToString(shift)
                 << ": ";

            for (int i = 0;
                 i < static_cast<int>(
                         schedule.assignments[day][shift].size());
                 i++)
            {
                int employeeIndex =
                    schedule.assignments[day][shift][i];

                cout << employees[employeeIndex].name;

                if (i <
                    static_cast<int>(
                        schedule.assignments[day][shift].size()) - 1)
                {
                    cout << ", ";
                }
            }

            cout << "\n";
        }
    }
}


// ============================================================
// Display Employee Workload
// ============================================================

void displayWorkload(
    const vector<Employee>& employees)
{
    cout << "\n";
    cout << "============================================================\n";
    cout << "                  EMPLOYEE WORKLOAD\n";
    cout << "============================================================\n";

    cout << left
         << setw(25)
         << "Employee"
         << setw(15)
         << "Days Worked"
         << "Status\n";

    cout << "------------------------------------------------------------\n";

    for (const Employee& employee : employees)
    {
        cout << left
             << setw(25)
             << employee.name
             << setw(15)
             << employee.daysWorked;

        if (employee.daysWorked <=
            MAX_DAYS_PER_EMPLOYEE)
        {
            cout << "Within 5-Day Limit";
        }
        else
        {
            cout << "EXCEEDS LIMIT";
        }

        cout << "\n";
    }
}


// ============================================================
// Display Preference Conflicts
// ============================================================

void displayConflicts(
    const Schedule& schedule)
{
    cout << "\n";
    cout << "============================================================\n";
    cout << "               SHIFT CONFLICT HANDLING\n";
    cout << "============================================================\n";

    if (schedule.conflicts.empty())
    {
        cout << "No preference conflicts occurred.\n";
    }
    else
    {
        for (const string& conflict :
             schedule.conflicts)
        {
            cout << "- " << conflict << "\n";
        }

        cout << "\n";
        cout << "Conflicts were resolved by assigning employees\n";
        cout << "to another available shift on the same day.\n";
    }
}


// ============================================================
// Display Validation Results
// ============================================================

void displayValidation(
    const Schedule& schedule,
    const vector<Employee>& employees)
{
    bool staffingPassed =
        validateMinimumStaffing(schedule);

    bool dailyPassed =
        validateOneShiftPerDay(
            schedule,
            employees.size()
        );

    bool workloadPassed =
        validateMaximumFiveDays(employees);

    cout << "\n";
    cout << "============================================================\n";
    cout << "                SCHEDULING VALIDATION\n";
    cout << "============================================================\n";

    cout << "\n";

    if (staffingPassed)
    {
        cout << "[PASS] Minimum 2 employees per shift\n";
    }
    else
    {
        cout << "[FAIL] Minimum 2 employees per shift\n";
    }

    if (dailyPassed)
    {
        cout << "[PASS] Maximum 1 shift per employee per day\n";
    }
    else
    {
        cout << "[FAIL] Maximum 1 shift per employee per day\n";
    }

    if (workloadPassed)
    {
        cout << "[PASS] Maximum 5 working days per employee\n";
    }
    else
    {
        cout << "[FAIL] Maximum 5 working days per employee\n";
    }

    if (!schedule.conflicts.empty())
    {
        cout << "[PASS] Shift conflict detection and resolution\n";
    }
    else
    {
        cout << "[PASS] Shift preference handling\n";
    }

    cout << "\n";
}


// ============================================================
// Display Program Summary
// ============================================================

void displaySummary(
    const vector<Employee>& employees)
{
    int totalAssignments = 0;

    for (const Employee& employee : employees)
    {
        totalAssignments += employee.daysWorked;
    }

    cout << "============================================================\n";
    cout << "                  SCHEDULE SUMMARY\n";
    cout << "============================================================\n";

    cout << "Employees:              "
         << employees.size() << "\n";

    cout << "Days:                   "
         << DAYS << "\n";

    cout << "Shifts per day:         "
         << SHIFTS << "\n";

    cout << "Employees per shift:    "
         << MIN_PER_SHIFT << "\n";

    cout << "Total assignments:      "
         << totalAssignments << "\n";

    cout << "Maximum days/employee:  "
         << MAX_DAYS_PER_EMPLOYEE << "\n";

    cout << "\n";
}


// ============================================================
// Main Function
// ============================================================

int main()
{
    cout << "\n";
    cout << "============================================================\n";
    cout << "             EMPLOYEE SHIFT SCHEDULER\n";
    cout << "             ASSIGNMENT 4 - C++ VERSION\n";
    cout << "============================================================\n";

    cout << "\n";
    cout << "Scheduling requirements:\n";
    cout << "  * 7 days per week\n";
    cout << "  * Morning, Afternoon, and Evening shifts\n";
    cout << "  * Minimum 2 employees per shift\n";
    cout << "  * Maximum 1 shift per employee per day\n";
    cout << "  * Maximum 5 working days per employee\n";
    cout << "  * Employee shift preferences are considered\n";
    cout << "  * Shift conflicts are detected and resolved\n";

    // --------------------------------------------------------
    // Input
    // --------------------------------------------------------

    vector<Employee> employees;

    inputEmployees(employees);

    // --------------------------------------------------------
    // Generate Schedule
    // --------------------------------------------------------

    cout << "\n";
    cout << "Generating weekly schedule...\n";

    Schedule schedule =
        generateSchedule(employees);

    // --------------------------------------------------------
    // Display Results
    // --------------------------------------------------------

    displaySummary(employees);

    displaySchedule(
        schedule,
        employees
    );

    displayWorkload(employees);

    displayConflicts(schedule);

    displayValidation(
        schedule,
        employees
    );

    // --------------------------------------------------------
    // Final Status
    // --------------------------------------------------------

    bool staffingPassed =
        validateMinimumStaffing(schedule);

    bool dailyPassed =
        validateOneShiftPerDay(
            schedule,
            employees.size()
        );

    bool workloadPassed =
        validateMaximumFiveDays(employees);

    cout << "\n";
    cout << "============================================================\n";

    if (staffingPassed &&
        dailyPassed &&
        workloadPassed)
    {
        cout << "           SCHEDULE GENERATED SUCCESSFULLY\n";
    }
    else
    {
        cout << "           SCHEDULE VALIDATION FAILED\n";
    }

    cout << "============================================================\n";

    cout << "\nThank you for using the Employee Shift Scheduler.\n";

    return 0;
}

File: employee_scheduler.py
from flask import Flask, request, render_template_string
import random

app = Flask(__name__)

DAYS = [
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday",
    "Sunday"
]

SHIFTS = ["Morning", "Afternoon", "Evening"]

MIN_EMPLOYEES = 9
MIN_PER_SHIFT = 2
MAX_DAYS_PER_EMPLOYEE = 5


HTML = """
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Employee Shift Scheduler</title>

    <style>
        * {
            box-sizing: border-box;
        }

        body {
            margin: 0;
            font-family: Arial, sans-serif;
            background: #f4f7fb;
            color: #1f2937;
        }

        .header {
            background: linear-gradient(135deg, #1e3a8a, #2563eb);
            color: white;
            padding: 35px 20px;
            text-align: center;
        }

        .header h1 {
            margin: 0 0 10px;
            font-size: 34px;
        }

        .header p {
            margin: 0;
            font-size: 16px;
        }

        .container {
            width: 94%;
            max-width: 1400px;
            margin: 30px auto;
        }

        .card {
            background: white;
            border-radius: 14px;
            padding: 25px;
            margin-bottom: 25px;
            box-shadow: 0 4px 15px rgba(0, 0, 0, 0.08);
        }

        .card h2 {
            margin-top: 0;
            color: #1e3a8a;
        }

        .notice {
            background: #eff6ff;
            border-left: 5px solid #2563eb;
            padding: 15px;
            margin-bottom: 20px;
            border-radius: 6px;
        }

        .employee-grid {
            overflow-x: auto;
        }

        table {
            width: 100%;
            border-collapse: collapse;
            min-width: 900px;
        }

        th {
            background: #1e3a8a;
            color: white;
            padding: 12px;
            text-align: center;
        }

        td {
            padding: 10px;
            border-bottom: 1px solid #e5e7eb;
            text-align: center;
        }

        td:first-child {
            font-weight: bold;
            text-align: left;
        }

        input[type="text"],
        select {
            width: 100%;
            padding: 9px;
            border: 1px solid #cbd5e1;
            border-radius: 6px;
            background: white;
        }

        .button-area {
            text-align: center;
            margin-top: 25px;
        }

        button {
            border: none;
            padding: 13px 25px;
            border-radius: 8px;
            font-size: 16px;
            font-weight: bold;
            cursor: pointer;
            margin: 5px;
        }

        .generate {
            background: #2563eb;
            color: white;
        }

        .generate:hover {
            background: #1d4ed8;
        }

        .reset {
            background: #e5e7eb;
            color: #1f2937;
        }

        .error {
            background: #fee2e2;
            border-left: 5px solid #dc2626;
            color: #991b1b;
            padding: 15px;
            margin-bottom: 20px;
            border-radius: 6px;
        }

        .success {
            background: #dcfce7;
            border-left: 5px solid #16a34a;
            color: #166534;
            padding: 15px;
            margin-bottom: 20px;
            border-radius: 6px;
        }

        .schedule-table {
            overflow-x: auto;
        }

        .schedule-table table {
            min-width: 1000px;
        }

        .schedule-table td {
            vertical-align: top;
            min-width: 200px;
            height: 100px;
        }

        .employee {
            display: block;
            background: #eff6ff;
            border-radius: 6px;
            padding: 7px;
            margin: 4px 0;
            font-weight: bold;
        }

        .day-name {
            background: #dbeafe;
            color: #1e3a8a;
            font-weight: bold;
            font-size: 16px;
        }

        .stats {
            display: flex;
            flex-wrap: wrap;
            gap: 15px;
            margin-bottom: 20px;
        }

        .stat {
            flex: 1;
            min-width: 180px;
            background: #f8fafc;
            padding: 18px;
            border-radius: 10px;
            text-align: center;
        }

        .stat-number {
            font-size: 28px;
            font-weight: bold;
            color: #2563eb;
        }

        .footer {
            text-align: center;
            padding: 25px;
            color: #64748b;
            font-size: 14px;
        }

        @media (max-width: 700px) {
            .header h1 {
                font-size: 26px;
            }

            .container {
                width: 98%;
            }

            .card {
                padding: 15px;
            }
        }
    </style>
</head>

<body>

<div class="header">
    <h1>Employee Shift Scheduler</h1>
    <p>Weekly employee scheduling with shift preferences and scheduling constraints</p>
</div>

<div class="container">

    {% if error %}
        <div class="error">
            <strong>Error:</strong> {{ error }}
        </div>
    {% endif %}

    {% if schedule %}
        <div class="success">
            <strong>Schedule generated successfully!</strong>
            All shifts contain at least two employees, no employee works more
            than one shift per day, and no employee works more than five days.
        </div>

        <div class="card">
            <h2>Schedule Summary</h2>

            <div class="stats">
                <div class="stat">
                    <div class="stat-number">{{ employee_count }}</div>
                    <div>Employees</div>
                </div>

                <div class="stat">
                    <div class="stat-number">7</div>
                    <div>Days</div>
                </div>

                <div class="stat">
                    <div class="stat-number">3</div>
                    <div>Shifts per Day</div>
                </div>

                <div class="stat">
                    <div class="stat-number">42</div>
                    <div>Total Assignments</div>
                </div>
            </div>
        </div>

        <div class="card">
            <h2>Final Weekly Schedule</h2>

            <div class="schedule-table">
                <table>
                    <tr>
                        <th>Day</th>
                        <th>Morning</th>
                        <th>Afternoon</th>
                        <th>Evening</th>
                    </tr>

                    {% for day in days %}
                    <tr>
                        <td class="day-name">{{ day }}</td>

                        {% for shift in shifts %}
                        <td>
                            {% for employee in schedule[day][shift] %}
                                <span class="employee">{{ employee }}</span>
                            {% endfor %}
                        </td>
                        {% endfor %}
                    </tr>
                    {% endfor %}
                </table>
            </div>
        </div>

        <div class="card">
            <h2>Employee Workload</h2>

            <div class="schedule-table">
                <table>
                    <tr>
                        <th>Employee</th>
                        <th>Days Worked</th>
                        <th>Status</th>
                    </tr>

                    {% for employee, count in workload.items() %}
                    <tr>
                        <td>{{ employee }}</td>
                        <td>{{ count }}</td>
                        <td>
                            {% if count <= 5 %}
                                Within 5-Day Limit
                            {% else %}
                                Exceeds Limit
                            {% endif %}
                        </td>
                    </tr>
                    {% endfor %}
                </table>
            </div>
        </div>

        <div class="card">
            <h2>Scheduling Notes</h2>
            <ul>
                <li>Each shift has at least two employees.</li>
                <li>An employee can work only one shift per day.</li>
                <li>An employee can work a maximum of five days per week.</li>
                <li>Employee preferences are considered during assignment.</li>
                <li>If a preferred shift is full, another available shift is used.</li>
                <li>Random selection is used when multiple employees have equal priority.</li>
            </ul>
        </div>

        <div class="button-area">
            <a href="/">
                <button class="reset">Create Another Schedule</button>
            </a>
        </div>

    {% else %}

        <div class="card">
            <h2>Enter Employee Information</h2>

            <div class="notice">
                <strong>Important:</strong>
                At least 9 employees are required for a complete weekly schedule.
                The company requires 2 employees for each of 3 shifts across 7 days,
                while each employee can work a maximum of 5 days.
            </div>

            <form method="POST">

                <div class="employee-grid">
                    <table>
                        <tr>
                            <th>Employee Name</th>
                            {% for day in days %}
                                <th>{{ day }}</th>
                            {% endfor %}
                        </tr>

                        {% for i in range(12) %}
                        <tr>
                            <td>
                                <input
                                    type="text"
                                    name="name_{{ i }}"
                                    placeholder="Employee {{ i + 1 }}"
                                >
                            </td>

                            {% for day in days %}
                            <td>
                                <select name="pref_{{ i }}_{{ loop.index0 }}">
                                    <option value="">No Preference</option>
                                    <option value="Morning">Morning</option>
                                    <option value="Afternoon">Afternoon</option>
                                    <option value="Evening">Evening</option>
                                </select>
                            </td>
                            {% endfor %}
                        </tr>
                        {% endfor %}
                    </table>
                </div>

                <div class="button-area">
                    <button type="submit" class="generate">
                        Generate Weekly Schedule
                    </button>
                </div>

            </form>
        </div>

        <div class="card">
            <h2>Scheduling Rules</h2>

            <ul>
                <li>Company operates seven days per week.</li>
                <li>Available shifts are Morning, Afternoon, and Evening.</li>
                <li>Each employee works no more than one shift per day.</li>
                <li>Each employee works no more than five days per week.</li>
                <li>Each shift requires at least two employees per day.</li>
                <li>Additional employees are selected randomly when necessary.</li>
                <li>Shift conflicts are handled by assigning another available shift.</li>
            </ul>
        </div>

    {% endif %}

</div>

<div class="footer">
    Assignment 4 - Implementing Control Structures | Python Employee Scheduler
</div>

</body>
</html>
"""


def create_schedule(employees):
    """
    Generate a weekly schedule.

    The algorithm uses:
    - conditionals
    - loops
    - branching
    - random selection
    - preference handling
    - workload constraints
    """

    schedule = {
        day: {
            shift: []
            for shift in SHIFTS
        }
        for day in DAYS
    }

    workload = {
        employee["name"]: 0
        for employee in employees
    }

    # Process each day.
    for day_index, day in enumerate(DAYS):

        # Employees already working today.
        working_today = set()

        # Randomize equal-priority choices.
        shuffled_employees = employees.copy()
        random.shuffle(shuffled_employees)

        # Assign each shift.
        for shift in SHIFTS:

            # Find employees who:
            # 1. have not worked today
            # 2. have not reached 5 days
            available = [
                employee
                for employee in shuffled_employees
                if employee["name"] not in working_today
                and workload[employee["name"]] < MAX_DAYS_PER_EMPLOYEE
            ]

            # Give preference to employees who requested this shift.
            preferred = [
                employee
                for employee in available
                if employee["preferences"][day_index] == shift
            ]

            # Sort candidates by workload.
            preferred.sort(
                key=lambda employee: workload[employee["name"]]
            )

            available.sort(
                key=lambda employee: workload[employee["name"]]
            )

            selected = []

            # First select employees who preferred this shift.
            for employee in preferred:
                if len(selected) >= MIN_PER_SHIFT:
                    break

                if employee["name"] not in working_today:
                    selected.append(employee)

            # If fewer than two preferred employees are available,
            # select other available employees.
            if len(selected) < MIN_PER_SHIFT:

                remaining = [
                    employee
                    for employee in available
                    if employee["name"] not in {
                        selected_employee["name"]
                        for selected_employee in selected
                    }
                ]

                # Randomize candidates with the same workload.
                random.shuffle(remaining)

                remaining.sort(
                    key=lambda employee: workload[employee["name"]]
                )

                for employee in remaining:
                    if len(selected) >= MIN_PER_SHIFT:
                        break

                    selected.append(employee)

            # Store assignments.
            for employee in selected:
                name = employee["name"]

                schedule[day][shift].append(name)
                workload[name] += 1
                working_today.add(name)

        # Verify that all three shifts have the required staffing.
        for shift in SHIFTS:
            if len(schedule[day][shift]) < MIN_PER_SHIFT:
                raise ValueError(
                    f"Unable to staff {day} {shift} shift."
                )

    return schedule, workload


@app.route("/", methods=["GET", "POST"])
def index():

    if request.method == "POST":

        employees = []
        names_seen = set()

        # Read employee information from the form.
        for i in range(12):

            name = request.form.get(
                f"name_{i}",
                ""
            ).strip()

            if not name:
                continue

            # Prevent duplicate employee names.
            if name.lower() in names_seen:
                return render_template_string(
                    HTML,
                    days=DAYS,
                    shifts=SHIFTS,
                    schedule=None,
                    error=f"Duplicate employee name: {name}",
                    workload=None,
                    employee_count=0
                )

            names_seen.add(name.lower())

            preferences = []

            for day_index in range(7):
                preference = request.form.get(
                    f"pref_{i}_{day_index}",
                    ""
                )

                preferences.append(preference)

            employees.append({
                "name": name,
                "preferences": preferences
            })

        # At least 9 employees are required.
        if len(employees) < MIN_EMPLOYEES:
            return render_template_string(
                HTML,
                days=DAYS,
                shifts=SHIFTS,
                schedule=None,
                error=(
                    f"Please enter at least {MIN_EMPLOYEES} employees. "
                    f"The current number is {len(employees)}."
                ),
                workload=None,
                employee_count=len(employees)
            )

        try:
            schedule, workload = create_schedule(employees)

        except ValueError as error:
            return render_template_string(
                HTML,
                days=DAYS,
                shifts=SHIFTS,
                schedule=None,
                error=str(error),
                workload=None,
                employee_count=len(employees)
            )

        return render_template_string(
            HTML,
            days=DAYS,
            shifts=SHIFTS,
            schedule=schedule,
            workload=workload,
            error=None,
            employee_count=len(employees)
        )

    return render_template_string(
        HTML,
        days=DAYS,
        shifts=SHIFTS,
        schedule=None,
        error=None,
        workload=None,
        employee_count=0
    )


if __name__ == "__main__":
    print("=" * 60)
    print("EMPLOYEE SHIFT SCHEDULER")
    print("=" * 60)
    print("Starting Flask web application...")
    print("Open your browser and go to:")
    print("http://127.0.0.1:5000")
    print("=" * 60)

    app.run(
        debug=True,
        host="127.0.0.1",
        port=5000
    )